Welcome to PodiumASM's documentation!
=====================================


.. raw:: html

   <img src="https://raw.githubusercontent.com/thdurand4/PodiumASM/main/docs/source/_images/PodiumASM_logo.png" align="right" alt="podiumASM Logo">

Today, assembly a genome using long reads from either Oxford Nanopore Technologies or Pacific Biosciences is really powerful, in particular to solve repeats and structural variants, for prokaryotic as well as for eukaryotic genomes. Such technologies provide assemblies that are increased in contiguity and accuracy.

Due to the daily deluge of data sequences and the increasing number of released tools that are even updated every week, many species see having their genome assembled.

But when you have to make an assembly of your species and don't know which assembler is the best for your organism, PodiumASM can help you

PodiumASM is is an open-source, scalable, modulable and traceable snakemake pipeline, able to compare multiple long read assemblies obtained from multiple assemblers tools. The workflow PodiumASM can help you to choose the best assemblies among all possibilities.


.. toctree::
   :caption: About PodiumASM
   :name: about_podium_asm
   :maxdepth: 2

   ABOUT.rst

.. toctree::
   :caption: Install
   :name: install
   :maxdepth: 2

   INSTALL.rst

.. toctree::
   :caption: Defining Workflows
   :name: defining_workflows
   :maxdepth: 2

   WORKFLOWS.rst

.. toctree::
   :caption: Project Info
   :name: project_info
   :maxdepth: 1

   PROJECT.rst

.. toctree::
   :caption: API
   :name: api_podiumASM
   :maxdepth: 2

   click.rst

Indices and tables
==================

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
